cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java

```diff
@@ -0,0 +1,197 @@
+    Set<ProgramId> servicesRunning = new HashSet<>();
+    Map<ProgramId, List<ProgramRuntimeService.RuntimeInfo>> programRunsMap = new HashMap<>();
+    Map<ProgramId, RunRecordDetail> programLatestRunMap = new HashMap<>();
+    //Get all current runs
+    List<ProgramRuntimeService.RuntimeInfo> runtimeInfos = programRuntimeService.listAll(ProgramType.SERVICE);
+    //Find services to delete and already running services
+    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
+      ProgramId programId = runtimeInfo.getProgramId();
+      if (!programId.getNamespaceId().equals(NamespaceId.SYSTEM)) {
+        //We care only about system services
+        continue;
+      }
+      if (!enabledServices.contains(programId)) {
+        servicesToStop.add(programId);
+      } else {
+        servicesRunning.add(programId);
+        //gather run records to stop extra runs for each program
+        gatherRunRecords(programRunsMap, programLatestRunMap, runtimeInfo, programId);
+      }
+    }
+    //find services to start
+    Set<ProgramId> servicesToStart = enabledServices.stream()
+      .filter(programId -> !servicesRunning.contains(programId))
+      .collect(Collectors.toSet());
+    //start services
+    startServices(servicesToStart, enabledServicesMap);
+    //stop services
+    servicesToStop.forEach(this::stopService);
+    //stop extra runs for enabled services
+    pruneRuns(programRunsMap, programLatestRunMap);
+  }
+
+  private void gatherRunRecords(Map<ProgramId, List<ProgramRuntimeService.RuntimeInfo>> programRunsMap,
+                                Map<ProgramId, RunRecordDetail> programLastRunMap,
+                                ProgramRuntimeService.RuntimeInfo runtimeInfo, ProgramId programId) throws Exception {
+    programRunsMap.putIfAbsent(programId, new ArrayList<>());
+    programRunsMap.get(programId).add(runtimeInfo);
+    //Gather the latest run as we go
+    RunRecordDetail runRecordMeta = programLifecycleService
+      .getRunRecordMeta(runtimeInfo.getController().getProgramRunId());
+    programLastRunMap.putIfAbsent(programId, runRecordMeta);
+    programLastRunMap.put(programId, getLatestRunRecord(programLastRunMap.get(programId), runRecordMeta));
+  }
+
+  private void startServices(Set<ProgramId> servicesToStart, Map<ProgramId, Map<String, String>> enabledServicesMap) {
+    for (ProgramId programId : servicesToStart) {
+      Map<String, String> overrides = enabledServicesMap.get(programId);
+      LOG.debug("Starting program {} with args {}", programId, overrides);
+      try {
+        programLifecycleService.start(programId, overrides, false);
+      } catch (Exception ex) {
+        LOG.error("Could not start program", ex);
+      }
+    }
+  }
+
+  private void stopService(ProgramId programId) {
+    LOG.debug("Stopping program {} ", programId);
+    try {
+      programLifecycleService.stop(programId);
+    } catch (Exception ex) {
+      LOG.error("Could not stop program", ex);
```

Comment:
Please log with more context. E.g. `"LOG.warn(Could not stop program {}. Will be retried", programId, ex)`.
Also, it should be logged as warning since it is not a fatal error and it will get retried.

Reply:
Done